pyomo/repn/plugins/gams_writer.py

```diff
@@ -0,0 +1,512 @@
+        put_results = io_options.pop("put_results", None)
+
+        if len(io_options):
+            raise ValueError(
+                "ProblemWriter_gams passed unrecognized io_options:\n\t" +
+                "\n\t".join("%s = %s"
+                            % (k,v) for k,v in io_options.iteritems()))
+
+        if solver is not None:
+            if solver.upper() not in valid_solvers:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "solver: %s" % solver)
+
+        if mtype is not None:
+            valid_mtypes = set([
+                'lp', 'qcp', 'nlp', 'dnlp', 'rmip', 'mip', 'rmiqcp', 'rminlp',
+                'miqcp', 'minlp', 'rmpec', 'mpec', 'mcp', 'cns', 'emp'])
+            if mtype.lower() not in valid_mtypes:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "model type: %s" % mtype)
+            if (solver is not None and
+                mtype.upper() not in valid_solvers[solver.upper()]):
+                raise ValueError("ProblemWriter_gams passed solver (%s) "
+                                 "unsuitable for given model type (%s)"
+                                 % (solver, mtype))
+
+        if output_filename is None:
+            output_filename = model.name + ".gms"
+
+        if symbolic_solver_labels and (labeler is not None):
+            raise ValueError("ProblemWriter_gams: Using both the "
+                             "'symbolic_solver_labels' and 'labeler' "
+                             "I/O options is forbidden")
+
+        if symbolic_solver_labels:
+            var_labeler = con_labeler = AlphaNumericTextLabeler()
+        elif labeler is None:
+            var_labeler = NumericLabeler('x')
+            con_labeler = NumericLabeler('c')
+        else:
+            var_labeler = con_labeler = labeler
+
+        var_list = []
+        symbolMap = SymbolMap()
+
+        def var_recorder(obj):
+            ans = var_labeler(obj)
+            var_list.append(ans)
+            return ans
+
+        def var_label(obj):
+            if obj.is_fixed():
+                return str(value(obj))
+            return symbolMap.getSymbol(obj, var_recorder)
+
+        # when sorting, there are a non-trivial number of
+        # temporary objects created. these all yield
+        # non-circular references, so disable GC - the
+        # overhead is non-trivial, and because references
+        # are non-circular, everything will be collected
+        # immediately anyway.
+        with PauseGC() as pgc:
+            try:
+                if isinstance(output_filename, string_types):
+                    output_file = open(output_filename, "w")
+                else:
+                    # Support passing of stream such as a StringIO
+                    # on which to write the model file
+                    output_file = output_filename
+                if isinstance(model, IBlockStorage):
```

Comment:
I've started implementing the to_string() keyword choice, but I am a little concerned. I started by updating all the to_string definitions that I knew of, both adding a labeler keyword to their definition and passing this labeler as a keyword to all calls it makes, but of course that was not all of the to_string methods spread around pyomo. Now I'm relying on the regression tests to tell me when I missed one. For example, I started in expr_coopr3 and did all of those, but then a baron test errored because it tried passing labeler to a SimpleExpression's to_string, so then I updated that one. On the other hand, I got an error because I was trying to pass labeler to a NumericConstant's to_string (part of the constant objective tests), but I decided that should just return a string of the value, not a label, so I changed my code to not pass a labeler if objective.is_constant(). With all of this, I worry that either I or the regression tests will miss a to_string() to change. Thoughts?

The nice thing about the ComponentData.labeler strategy was that nothing had to be passed, and the appropriate __str__ or to_string functions knew whether or not to call this labeler. But now it needs to be passed quite a bit.

As a second-hand thought, should every single to_string method definition in pyomo be updated to include the labeler keyword, and only the appropriate (ComponentData) objects be told to use it?